### PR TITLE
Clean up the internal APIs used for stack trace construction.

### PR DESCRIPTION
**Pull request by sjrd**
Prompted by #4710.

--

If #4721 is merged before this PR, we should also change the code in `javalibintf/StackTraceElement.scala` to use the new constructor of `java.lang.StackTraceElement`.

### CODE REVIEW

**Comment by gzm0** [comment]
QQ I had when I saw this: Do we even still need the column number in `StackTraceElement`? IIUC, it was there to allow for post stack trace conversion source map processing. At least in the Scala.js codebase, it only seems to be used in tests.

Maybe we should just remove it? The last commit makes it clear that additional fields for data-only classes open up all kinds of cans of worms.

**Comment by sjrd** [comment]
I believe we should keep it, yes.

It's used in `toString`, which means it's displayed in stack trace dumps. That's pretty important.

The JavaScript world in general standardizes on having column numbers part of stack traces. It doesn't seem like a good idea to me to get rid of them. Even if we don't process them, it's not unlikely that some other tool dealing with stack traces would. In fact, I think @sherpal recently built an in-browser stack trace display that directly reads the column number.

**Comment by sherpal** [comment]
Ah yes that's right. Without the column number, it would be impossible to do that. And I believe it's a big plus to be able to display properly stack traces like it does.

File: javalib/src/main/scala/java/lang/StackTrace.scala [hunk 1]
```diff
@@ -105,8 +105,7 @@ private[lang] object StackTrace {
   private def normalizedLinesToStackTrace(
       lines: js.Array[String]): Array[StackTraceElement] = {
 
-    val NormalizedFrameLine = """^([^\@]*)\@(.*):([0-9]+)$""".re
-    val NormalizedFrameLineWithColumn = """^([^\@]*)\@(.*):([0-9]+):([0-9]+)$""".re
+    val NormalizedFrameLine = """^([^@]*)@(.*?):([0-9]+)(?::([0-9]+))?$""".re
 
     @inline def parseInt(s: String): Int =
       js.Dynamic.global.parseInt(s).asInstanceOf[Int]
```
Comment (gzm0) [inline javalib/src/main/scala/java/lang/StackTrace.scala:111]: Somewhat unrelated, but shouldn't this set the radix explicitly?

File: javalib/src/main/scala/java/lang/StackTrace.scala [hunk 2]
```diff
@@ -105,8 +105,7 @@ private[lang] object StackTrace {
   private def normalizedLinesToStackTrace(
       lines: js.Array[String]): Array[StackTraceElement] = {
 
-    val NormalizedFrameLine = """^([^\@]*)\@(.*):([0-9]+)$""".re
-    val NormalizedFrameLineWithColumn = """^([^\@]*)\@(.*):([0-9]+):([0-9]+)$""".re
+    val NormalizedFrameLine = """^([^@]*)@(.*?):([0-9]+)(?::([0-9]+))?$""".re
```
Comment (gzm0) [inline javalib/src/main/scala/java/lang/StackTrace.scala:108]: Merging these two regexes seems like a good idea, but maybe it should go into a separate commit?

**Review by gzm0 — CHANGES_REQUESTED** [review]
Mostly nits. Potential contentious point: I don't think we should leave the unused deprecated methods around.

File: javalib/src/main/scala/java/lang/Throwables.scala [hunk 3]
```diff
@@ -45,26 +45,28 @@ class Throwable protected (s: String, private var e: Throwable,
...
    */
+  @deprecated("old internal API", "1.11.0")
   def getStackTraceStateInternal(): Any =
-    stackTraceStateInternal
+    jsErrorForStackTrace
 
   /* Not part of the JDK API, used internally in java.lang and accessible
    * through reflection.
    */
+  @deprecated("old internal API", "1.11.0")
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/lang/Throwables.scala:62]: This one I think we can remove. It's clearly *named* Internal, and we have no public tests for it.

File: javalib/src/main/scala/java/lang/StackTraceElement.scala [hunk 4]
```diff
@@ -15,25 +15,28 @@ package java.lang
...
-   * through reflection.
-   */
+  // Not part of the JDK API, accessible through reflection.
   def getColumnNumber(): Int = columnNumber
 
-  /* Not part of the JDK API, used internally in java.lang and accessible
-   * through reflection.
-   */
+  // Not part of the JDK API, accessible through reflection.
+  @deprecated("old internal API; use the constructor with a column number instead", "1.11.0")
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/lang/StackTraceElement.scala:39]: This one I think we should keep. I did not find any online evidence that it is used, but it does have public tests. It's also the only way that some stack processing code could have created `StackTraceElement`s with column numbers, so it probably has legitimate use cases.

File: javalib/src/main/scala/java/lang/StackTrace.scala (hunk 1 above)
Comment (sjrd) [reply javalib/src/main/scala/java/lang/StackTrace.scala:111]: Not in this case. The strings come straight for a regex that only takes `[0-9]+`, so that will always be interpreted in base 10.

**Comment by sjrd** [comment]
Updated.

I extracted the two suggested separate commits, and I removed the `{get,set}StackTraceStateInternal` methods in `Throwable`. I did keep `STE.setColumnNumber` so far.

File: javalib/src/main/scala/java/lang/StackTrace.scala (hunk 1 above)
Comment (gzm0) [reply javalib/src/main/scala/java/lang/StackTrace.scala:111]: Hah, I ran into this trap: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#octal_interpretations_with_no_radix

**Comment by sjrd** [comment]
Rebased.

**Review by gzm0 — APPROVED** [review]
